proto/structured.proto

```diff
@@ -27,61 +27,51 @@ option (gogoproto.sizer_all) = true;
 option (gogoproto.marshaler_all) = true;
 option (gogoproto.unmarshaler_all) = true;
 
-message Table {
-  optional string name = 1 [(gogoproto.nullable) = false];
+// For Cockroach internal use.
+message Internal {
+  optional uint32 id = 1 [(gogoproto.nullable) = false];
 }
-
+ 
 message Column {
   enum ColumnType {
     BYTES = 0;
   }
-
   optional string name = 1 [(gogoproto.nullable) = false];
   optional ColumnType type = 2 [(gogoproto.nullable) = false];
+  // For Cockroach internal use.
+  optional Internal internal = 9 [(gogoproto.nullable) = true];
 }
 
 message Index {
   optional string name = 1 [(gogoproto.nullable) = false];
   optional bool unique = 2 [(gogoproto.nullable) = false];
+  // An ordered list of column names of which the index is comprised. Each
+  // column_name refers to a column in the TableSchema's columns. Special
+  // care is taken to update this when deleting columns.
+  repeated string column_names = 3 [(gogoproto.nullable) = false];
+  // For Cockroach internal use.
+  optional Internal internal = 9 [(gogoproto.nullable) = true];
```

Comment:
Pete,

This has nothing to do with safety. I'd like to have an internal section in
the protobuf to store internal data related to the entity.

It occurred to me that go composition exports the internal struct methods
to the outer interface. So we do not need to say GetInternal().Id(), but
just GetId() to get to the Id. I'll check this as soon as I get to a
computer.

I'll rework this PR and get back to you.

Vivek

On Fri, Jun 5, 2015, 4:58 PM Peter Mattis <notifications@github.com> wrote:

> In proto/structured.proto
> <https://github.com/cockroachdb/cockroach/pull/1278#discussion_r31850955>:
>
> >  }
> >
> >  message Index {
> >    optional string name = 1 [(gogoproto.nullable) = false];
> >    optional bool unique = 2 [(gogoproto.nullable) = false];
> > +  // An ordered list of column names of which the index is comprised. Each
> > +  // column_name refers to a column in the TableSchema's columns. Special
> > +  // care is taken to update this when deleting columns.
> > +  repeated string column_names = 3 [(gogoproto.nullable) = false];
> > +  // For Cockroach internal use.
> > +  optional Internal internal = 9 [(gogoproto.nullable) = true];
>
> I'm confused: it's a done deal that we're using an Internal message? I
> have other problems with it. I doubt you're going to want the same internal
> fields for Index, Column and TableSchema. The name Internal is much to
> general: something like SchemaInternal would be more appropriate.
>
> But back to the meat of the issue: if we're going to be putting ids into
> the schema protos, I think we should just name them id and not burying
> them in the message. If the ids are in the schema protos, we're going to
> need to check that they aren't being sent incorrectly from the client
> making it difficult to misuse them. I don't see the extra safety of placing
> them in an Internal message as providing additional safety.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/cockroachdb/cockroach/pull/1278/files#r31850955>.
>